### PR TITLE
indexed foldable-traversable instances

### PR DESCRIPTION
**Pull request by matthewleon**

### CODE REVIEW

**Comment by paf31** [comment]
Thanks!

Maybe we should deprecate `Data.List.mapWithIndex` now. Also, would you like to add `TraversableWithIndex` here too?

**Comment by matthewleon** [comment]
Not sure how deprecation works in these libs, but I've done the rest.

**Comment by paf31** [comment]
Thanks! I think a comment is enough for now.

**Comment by paf31** [comment]
The alternative is to use a `Warn` constraint.

**Comment by matthewleon** [comment]
Looks like I've made a test fail now. Will look into this.

**Comment by matthewleon** [comment]
That test passes on my machine. Any idea what's going on there?

**Comment by paf31** [comment]
It appears there is actually a bug:

```
> mapWithIndex (\x ix -> x + ix) (fromFoldable [0, 1, 2, 3])
(3 : 3 : 3 : 3 : Nil)
```

File: src/Data/List/Lazy/Types.purs [hunk 1]
```diff
@@ -105,6 +107,12 @@ instance functorList :: Functor List where
     go Nil = Nil
     go (Cons x xs') = Cons (f x) (f <$> xs')
 
+instance functorWithIndexList :: FunctorWithIndex Int List where
+  mapWithIndex f xs = List (go 0 <$> unwrap xs)
+    where
+    go _ Nil = Nil
+    go i (Cons x xs') = Cons (f i x) (f (i + 1) <$> xs')
```
Comment (paf31) [inline src/Data/List/Lazy/Types.purs]: Shouldn't there be a recursive call to `go` here?

Maybe it's better to try to write this as a `foldr`?

Comment (matthewleon) [reply src/Data/List/Lazy/Types.purs]: Ah, indeed. I think I was running my tests locally on the wrong branch by accident.

Comment (matthewleon) [reply src/Data/List/Lazy/Types.purs]: Actually I think this is a different mistake I've made, besides the one breaking the test above. I'll go through and add some more thorough testing wherever it makes sense.

**Comment by matthewleon** [comment]
Okay, this should be good to go now.

File: src/Data/List/Lazy/Types.purs [hunk 2]
```diff
@@ -120,6 +125,22 @@ instance foldableList :: Foldable List where
 
   foldMap f = foldl (\b a -> b <> f a) mempty
 
+instance foldableWithIndexList :: FoldableWithIndex Int List where
+  foldrWithIndex f b xs =
```
Comment (paf31) [inline src/Data/List/Lazy/Types.purs]: This seems like we could extract a default implementation in terms of `foldlWithIndex`. What do you think?

Comment (matthewleon) [reply src/Data/List/Lazy/Types.purs]: I'll try.

**Comment by matthewleon** [comment]
I don't think we can build a polymorphic default `foldrWithIndex` in terms of `foldlWithIndex`, actually. To begin, I don't think we can do a default `foldr` in terms of `foldl`.

My reasoning:

We know `foldr f a xs = foldl (flip f) a (reverse xs)`.

But this depends on a polymorphic `reverse`. 

There's one here for `Traversable`s: https://wiki.haskell.org/Foldable_and_Traversable
```haskell
reverseT :: (Traversable t) => t a -> t a
reverseT t = snd (mapAccumR (\ (x:xs) _ -> (xs, x)) (toList t) t)
```

But this, in turn, depends on a `toList`, and our library dependencies go in the other direction.

Now, we _do_ have a default implementation of `foldr` in terms of `foldMap`, which in turn is implemented in terms of `foldl`. But this uses the `Endo` monoid to build a thunk while traversing the `Foldable`, which kills our stack safety.

Please let me know if you think of something I've missed here. I would really love to get an efficient, stack-safe default implementation into `foldable-traversable`. But I can't think of any trick to do so.

**Comment by paf31** [comment]
Looks good, and I think you're right about the default implementations.

Could you please add a test for each of the new methods though?

**Comment by matthewleon** [comment]
Good idea. I will do that.

**Comment by matthewleon** [comment]
tests done

File: src/Data/List/Types.purs [hunk 3]
```diff
@@ -91,6 +110,11 @@ instance traversableList :: Traversable List where
   traverse f = map (foldl (flip (:)) Nil) <<< foldl (\acc -> lift2 (flip (:)) acc <<< f) (pure Nil)
   sequence = traverse id
 
+instance traversableWithIndexList :: TraversableWithIndex Int List where
+  traverseWithIndex f =
+    map (foldl (flip (:)) Nil)
```
Comment (paf31) [inline src/Data/List/Types.purs]: Isn't this `reverse`?

File: test/Test/Data/List.purs [hunk 4]
```diff
@@ -354,6 +375,13 @@ testList = do
   let xs = fromFoldable (range 1 100000)
   assert $ traverse Just xs == Just xs
 
+  log "traverseWithIndex should be stack-safe"
+  assert $ traverseWithIndex (const Just) xs == Just xs
+
+  log "traverseWithIndex should be correct"
+  assert $ traverseWithIndex (\i a -> Just $ i + a) (fromFoldable [2, 2, 2])
```
Comment (paf31) [inline test/Test/Data/List.purs]: Nitpicking, sorry: could you use an Applicative which tests that the order of effects is correct, like `Array`?

Comment (paf31) [reply test/Test/Data/List.purs]: Hmm, actually that wouldn't work, never mind.

File: src/Data/List/Types.purs (hunk 3 above)
Comment (matthewleon) [reply src/Data/List/Types.purs]: Yes, but we can't use it because of the reverse dependency.

**Comment by paf31** [comment]
Thanks!

**Pull request merged**